lib/AST/ClangTypeConverter.cpp

```diff
@@ -0,0 +1,684 @@
+  auto proto = type->getDecl();
+
+  // Single protocol -> id<Proto>
+  if (proto->isObjC()) {
+    auto &clangCtx = ClangASTContext;
+    clang::IdentifierInfo *name = &clangCtx.Idents.get(proto->getName().get());
+    auto *PDecl = clang::ObjCProtocolDecl::Create(
+                    const_cast<clang::ASTContext &>(clangCtx),
+                    clangCtx.getTranslationUnitDecl(), name,
+                    clang::SourceLocation(), clang::SourceLocation(), nullptr);
+
+    // Attach an objc_runtime_name attribute with the Objective-C name to use
+    // for this protocol.
+    SmallString<64> runtimeNameBuffer;
+    PDecl->addAttr(clang::ObjCRuntimeNameAttr::CreateImplicit(
+                     PDecl->getASTContext(),
+                     proto->getObjCRuntimeName(runtimeNameBuffer)));
+
+    auto clangType  = clangCtx.getObjCObjectType(clangCtx.ObjCBuiltinIdTy,
+                                                 &PDecl, 1);
+    return clangCtx.getObjCObjectPointerType(clangType);
+  }
+
+  return getClangIdType(ClangASTContext);
+}
+
+clang::QualType ClangTypeConverter::visitMetatypeType(MetatypeType *type) {
+  return getClangMetatypeType(ClangASTContext);
+}
+
+clang::QualType
+ClangTypeConverter::visitExistentialMetatypeType(ExistentialMetatypeType *type) {
+  return getClangMetatypeType(ClangASTContext);
+}
+
+clang::QualType ClangTypeConverter::visitClassType(ClassType *type) {
+  auto &clangCtx = ClangASTContext;
+  // produce the clang type INTF * if it is imported ObjC object.
+  auto swiftDecl = type->getDecl();
+  if (swiftDecl->isObjC()) {
+    clang::IdentifierInfo *ForwardClassId =
+      &clangCtx.Idents.get(swiftDecl->getName().get());
+    auto *CDecl = clang::ObjCInterfaceDecl::Create(
+                          clangCtx, clangCtx.getTranslationUnitDecl(),
+                          clang::SourceLocation(), ForwardClassId,
+                          /*typeParamList*/nullptr, /*PrevDecl=*/nullptr,
+                          clang::SourceLocation());
+
+    // Attach an objc_runtime_name attribute with the Objective-C name to use
+    // for this class.
+    SmallString<64> runtimeNameBuffer;
+    CDecl->addAttr(clang::ObjCRuntimeNameAttr::CreateImplicit(
+                     CDecl->getASTContext(),
+                     swiftDecl->getObjCRuntimeName(runtimeNameBuffer)));
+
+    auto clangType  = clangCtx.getObjCInterfaceType(CDecl);
+    return clangCtx.getObjCObjectPointerType(clangType);
+  }
+  return getClangIdType(clangCtx);
+}
+
+clang::QualType
+ClangTypeConverter::visitBoundGenericClassType(BoundGenericClassType *type) {
+  // Any @objc class type in Swift that shows up in an @objc method maps 1-1 to
+  // "id <SomeProto>"; with clang's encoding ignoring the protocol list.
+  return getClangIdType(ClangASTContext);
+}
+
+clang::QualType
+ClangTypeConverter::visitBoundGenericType(BoundGenericType *type) {
+  // The only possibilities are *Pointer<T>, SIMD*<T>,
+  // ImplicitlyUnwrappedOptional<T> and Optional<T>.
```

Comment:
This is another point where it's interesting to ask whether this function should be allowed to assume that the type is exportable to C as a precondition.  I'm almost positive that there are APIs for asking whether a type is any of these standard library types.

Reply:
> This is another point where it's interesting to ask whether this function should be allowed to assume that the type is exportable to C as a precondition.

Strictly speaking, it doesn't assume that (unless I made a mistake). If you look at the first item `switch (kind) { case StructKind::Invalid: `, in `GenClangType` we have a `llvm_unreachable` whereas in `ClangTypeConverter` we silently fail with `return clang::QualType()`.

This is important because someone might make a mistake in typing something and we should be issuing a diagnostic instead of crashing. I've replaced other similar crash points in `GenClangType` with `return clang::QualType()` as well.

Reply:
Hmm, okay.

I do think you probably at least need a bridged-to-ObjC check here.